### PR TITLE
[Merged by Bors] - refactor(NumberTheory.NumberField.CanonicalEmbedding): make the canonical embedding canonical

### PR DESCRIPTION
**Pull request by xroblot**
It was observed by @kbuzzard, and rightly so, that in the first version of this file the `canonical_embedding` was not so canonical. This refactor fixes that by replacing it with a truly canonical embedding. 

More precisely, in the old version, the canonical embedding was defined as the ring homomorphism
 `K →+* ℝ^r₁ × ℂ^r₂` that sends `x ∈ K` to `(φ_₁(x),...,φ_r₁(x)) × (ψ_₁(x),..., ψ_r₂(x))` where
 `φ_₁,...,φ_r₁` are its real embeddings and `ψ_₁,..., ψ_r₂` are its complex embeddings (up to
 complex conjugation). This is not canonical since it depends upon the choice of the `ψ`'s. 

In the new version, the canonical embedding is defined as the ring homomorphism `K →+* ℂ^n` that sends `x ∈ K` to `(φ_₁(x),...,φ_n(x))` where the `φ_i`'s are the complex embeddings of `K`.

The new version is easier to compute with since one does not have to distinguish between real and complex embeddings as in the first version. It also enables to prove the following result: the image of the ring of integers by `canonicalEmbedding` is a full ℤ-lattice (see `latticeBasis`) that will be useful in #5650.

Note that the original version of the canonical embedding will reappear in #5650 (as `mixedEmbedding`) since it allows for easier volume computation. 

---

- [x] depends on: #5996 
- [x] depends on: #6164 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
Please remember to add the label `awaiting-review` when this is ready for review. (Alternatively PRs that refactor files that were ported from mathlib3 should be labelled `after-port`.)

**Comment by xroblot** [comment]
> sorry, can you briefly explain what is the issue with the original one and how this fixes this?

Thanks for the suggestion. I have expanded the PR description with more details on the refactor.

File: Mathlib/NumberTheory/NumberField/CanonicalEmbedding.lean [hunk 1]
```diff
@@ -4,16 +4,16 @@ Released under Apache 2.0 license as described in the file LICENSE.
...
 /-!
 # Canonical embedding of a number field
 
-The canonical embedding of a number field `K` of signature `(r₁, r₂)` is the ring homomorphism
-`K →+* ℝ^r₁ × ℂ^r₂` that sends `x ∈ K` to `(φ_₁(x),...,φ_r₁(x)) × (ψ_₁(x),..., ψ_r₂(x))` where
-`φ_₁,...,φ_r₁` are its real embeddings and `ψ_₁,..., ψ_r₂` are its complex embeddings (up to
-complex conjugation).
+The canonical embedding of a number field `K` of degree `n` is the ring homomorphism
+`K →+* ℂ^n` that sends `x ∈ K` to `(φ_₁(x),...,φ_n(x))` where the `φ_i`'s are the complex
+embeddings of `K`.
```
Comment (kbuzzard) [inline Mathlib/NumberTheory/NumberField/CanonicalEmbedding.lean:16]:
```suggestion
embeddings of `K`. In fact we do not choose an ordering of the embeddings, mapping instead to the type of complex-valued functions on the type of field homomorphisms `K →+* ℂ`, a type of size `n`.
```

Comment (kbuzzard) [inline Mathlib/NumberTheory/NumberField/CanonicalEmbedding.lean:16]: You should add the canonical embedding under `## Main definitions and results`.

**Review by kbuzzard — COMMENTED** [review]
I haven't reviewed most of the later proofs, but I do think that this is the way to do it. The "canonical" R-vector space is the Galois-invariant subset of the C-vector-space. I have expressed my distaste for is-R-or-C before -- PRs like this are somehow doing things in a slightly more invariant way -- you get more theorems about complex conjugation for free if you consider it as an action of Z/2 on the RBarification rather than choosing an orientation (i.e. choosing `I` as a basis vector) and then having to start fussing about "CM types" or whatever they're called in the CM theory of abelian varieties.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**